### ISSUE TITLE
Idea: use `go run` in generate line

### ISSUE THREAD

**Issue by zombiezen** [google/wire#160]
[An idea](https://github.com/google/wire/issues/106#issuecomment-472934878) suggested by @steebchen is to use `//go:generate go run github.com/google/wire/cmd/wire` as the automatically inserted generate line for Wire output. This would have the advantage of always using a version of the code generator that's the same as the library being used, at the cost of perhaps some running time.

**EDIT** (Hit send too early) I don't think we want to make this change until Go Modules is the default, thus I'm marking this issue blocked.

**Comment by ernado** [comment]
This probably should be blocked until https://github.com/golang/go/issues/33468 is resolved.

Also, we can detect case when binary is started with `go run`, so there will be no backward compatibility issues at all. That is, if wired binary was started via `go run` we emit `go:generate go run github.com/google/wire/cmd/wire`, otherwise we fallback to current `go:generate wired`.

**Comment by steebchen** [comment]
Go modules is pretty much the default now, so that part should not be blocking anymore.

**Comment by steebchen** [comment]
I want to raise this again – go modules is the way to go by now.

**Comment by zombiezen** [comment]
With Go 1.16 having `GO111MODULE=on` by default, I agree. Happy to accept a PR to fix.

**Comment by steebchen** [comment]
Sounds good, I created #268

**Comment by odsod** [comment]
@zombiezen This is something we've been looking for too, to simplify our `go generate` invocations (we currently pre-build the correct version of the `wire` CLI and add it to `$PATH` before invoking `go generate`.

When this gets merged to master, we'd be happy to test this out and provide feedback before the next stable release.

**Comment by ernado** [comment]
Note that `go run` will be slower that direct call of `wire` because `go run` does not cache linked binary.

Also just simply changing from `wire` to `go run` can break some users, I've implemented backward compatible version in #239

**Comment by steebchen** [comment]
> Note that go run will be slower that direct call of wire because go run does not cache linked binary.

That's negligible:

```
❯ time go run github.com/google/wire/cmd/wire
wire: api: wrote /Users/luca/projects/ws-api/wire_gen.go
go run github.com/google/wire/cmd/wire  5.27s user 1.60s system 274% cpu 2.503 total

❯ time wire
wire: api: wrote /Users/luca/projects/ws-api/wire_gen.go
wire  5.26s user 1.05s system 236% cpu 2.673 total
```

> Also just simply changing from wire to go run can break some users, I've implemented backward compatible version in #239

Good point. However, I think either way the new default should be to use go generate, so that new users can make immediate use of it. Users who explicitly want the old behaviour should not upgrade, or set the option explicitly should they update.
